crates/synthesis/src/qsd.rs

```diff
@@ -0,0 +1,800 @@
+use std::sync::OnceLock;
+
+use approx::abs_diff_eq;
+use hashbrown::HashMap;
+use nalgebra::{stack, DMatrix, DMatrixView, DVector, Matrix4, QR};
+use ndarray::prelude::*;
+use num_complex::Complex64;
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use smallvec::smallvec;
+
+use crate::euler_one_qubit_decomposer::{
+    unitary_to_gate_sequence_inner, EulerBasis, EulerBasisSet,
+};
+use crate::linalg::{closest_unitary, is_hermitian_matrix, svd_decomposition, verify_unitary};
+use crate::two_qubit_decompose::{two_qubit_decompose_up_to_diagonal, TwoQubitBasisDecomposer};
+use qiskit_circuit::bit::ShareableQubit;
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::interner::Interned;
+use qiskit_circuit::operations::{ArrayType, OperationRef, Param, StandardGate, UnitaryGate};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+use qiskit_circuit::{Qubit, VarsMode};
+use qiskit_quantum_info::convert_2q_block_matrix::instructions_to_matrix;
+
+const EPS: f64 = 1e-10;
+
+enum VWType {
+    All,
+    OnlyV,
+    OnlyW,
+}
+
+pub fn quantum_shannon_decomposition(
+    mat: &DMatrix<Complex64>,
+    opt_a1: Option<bool>,
+    opt_a2: Option<bool>,
+    two_qubit_decomposer: Option<&TwoQubitBasisDecomposer>,
+    one_qubit_decomposer: Option<&EulerBasisSet>,
+) -> PyResult<CircuitData> {
+    let dim = mat.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+
+    if abs_diff_eq!(DMatrix::identity(dim, dim), mat) {
+        let out_qubits = (0..num_qubits)
+            .map(|_| ShareableQubit::new_anonymous())
+            .collect::<Vec<_>>();
+        return CircuitData::new(Some(out_qubits), None, None, 0, Param::Float(0.));
+    }
+    qsd_inner(
+        mat,
+        opt_a1,
+        opt_a2,
+        two_qubit_decomposer,
+        one_qubit_decomposer,
+        0,
+    )
+}
+
+fn qsd_inner(
+    mat: &DMatrix<Complex64>,
+    opt_a1: Option<bool>,
+    opt_a2: Option<bool>,
+    two_qubit_decomposer: Option<&TwoQubitBasisDecomposer>,
+    one_qubit_decomposer: Option<&EulerBasisSet>,
+    depth: usize,
+) -> PyResult<CircuitData> {
+    let dim = mat.shape().0;
+    let num_qubits = dim.ilog2() as usize;
+    let opt_a1_val = opt_a1.unwrap_or(true);
+    let mut default_1q_basis = EulerBasisSet::new();
+    default_1q_basis.add_basis(EulerBasis::U);
+    let default_2q_decomposer = TwoQubitBasisDecomposer::new_inner(
+        StandardGate::CX.into(),
+        smallvec![],
+        aview2(&qiskit_circuit::gate_matrix::CX_GATE),
+        1.0,
+        "U",
+        None,
+    )?;
+    let one_qubit_decomposer = one_qubit_decomposer.unwrap_or(&default_1q_basis);
+    let two_qubit_decomposer = two_qubit_decomposer.unwrap_or(&default_2q_decomposer);
+    if dim == 2 {
+        let array = Array2::from_shape_fn((dim, dim), |(i, j)| mat[(i, j)]);
+        let sequence =
+            unitary_to_gate_sequence_inner(array.view(), one_qubit_decomposer, 0, None, true, None);
```

Comment:
You can avoid the copy here with:

```suggestion
        let dim = ::ndarray::Dim(mat.shape());
        let strides = ::ndarray::Dim(mat.strides());
        // SAFETY: We know the array is a 2x2 and contiguous block (DMatrix uses a vec for backing storage) so we
        // don't need to check for invalid format
        let array = unsafe { ArrayView2::from_shape_ptr(dim.strides(strides), mat.get_unchecked(0)) };
        let sequence =
            unitary_to_gate_sequence_inner(array, one_qubit_decomposer, 0, None, true, None);
```